### ISSUE TITLE
line 21 in local.ts prevent sending request with the user data in the body.

### ISSUE THREAD

**Issue by yahyaKacem** [ronzeidman/ng2-ui-auth#19]
Hi, thanks for this lib, I'm trying to use for authentication but the line [21](https://github.com/ronzeidman/ng2-ui-auth/blob/master/src/local.ts#L21) in the local.ts file is preventing me from doing that my use case is to send a POST request to the back-end with the content-type of `application/x-www-form-urlencoded` and with the user data in the body, so I tried doing that by passing false instead of the user and the data in the second opts param, but the request was rejected because it was sent with false as params since the line 21 is:
```
opts.body = JSON.stringify(user) || opts.body;
``` 
so opts.body will be either the `JSON.stringify(user)` if the user is provided or `JSON.stringify(false)` which is not false it is the string "false" meaning the opts.body if it was passed it will never be used since `JSON.stringify(false)` is never false so to fix that:
```
opts.body = user ? JSON.stringify(user) : opts.body;
```
But, if I can give you a suggesting I would say make the whole function like this:
```
login(opts: any = {}, user: any = {}): Observable<Response> {
  let url: string;
  if (typeof opts === 'string') {
    url    = opts;
    opts = {};
  } else {
    url = opts.url ? opts.url : joinUrl(this.config.baseUrl, this.config.loginUrl);
  }
  opts.body = opts.body || JSON.stringify(user);
  opts.method = opts.method || 'POST';
  /*opts.headers = new Headers({
    'Content-Type': 'application/json'
  });*/
  return this.http.request(url, opts)
                  .map((response: Response) => {
                    this.shared.setToken(response);
                    return response;
                  });
}
```
this way you can call the login function in multiple cases without having to put false as the first parameter like this:
```
this.auth.login('/signin', user).subscribe((response) => this.goToMain(response));
// or
let options: RequestOptionsArgs = {
  method: 'POST',
  headers: headers,
  url: '/signin',
  body: user
};
this.auth.login(options).subscribe((response) => this.goToMain(response));
// or
let options: RequestOptionsArgs = {body: user};
this.auth.login(options).subscribe((response) => this.goToMain(response));
// or
let options: RequestOptionsArgs = {body: user, headers: headers};
this.auth.login('/signin', options).subscribe((response) => this.goToMain(response));
```

`**Note:** keep in mind that all the options passed to the login function are optional since they have default values.
`
Sorry for a long issue, and thanks again.

**Comment by ronzeidman** [comment]
Hi,
You're right, this line: `opts.body = JSON.stringify(user) || opts.body;` is meaningless...
Should be reversed.
Since I don't think introducing a breaking change is required here (that what will happen if I'll change the function signature) I think I'll fix it by writing this instead:
```ts
if (user) {
    opts.body = typeof user === 'string' ? user : JSON.stringify(user);
}
```
what do you say?

**Comment by ronzeidman** [comment]
notice that you can already define default headers: 
```ts
NG2_UI_AUTH_PROVIDERS({ 
    defaultHeaders: {'Content-Type': 'application/json'}, 
    loginUrl: '/signin',
    providers: {google: {clientId: GOOGLE_CLIENT_ID}}})
```
So you won't need to pass any options in your scenario.

**Comment by yahyaKacem** [comment]
OK, that make sense, I tried this and it worked except, I found a problem with the default headers, I get this error:
> Argument of type '{ authToken: string; baseUrl:
 string; loginUrl: string; tokenName: string; defaultHeaders: { 'Con...' is not assignable to parameter of type 'ICustomConfig'.
  Types of property 'defaultHeaders' are incompatible.
    Type '{ 'Content-Type': string; }' is not assignable to type '{ [name: string]: string; }'.
      Index signature is missing in type '{ 'Content-Type': string; }'.

I found out that it's a [bug with the current typescript compiler](http://stackoverflow.com/questions/22077023/why-cant-i-indirectly-return-an-object-literal-to-satisfy-an-index-signature-re), so to get around that instead of doing this:
```
NG2_UI_AUTH_PROVIDERS({defaultHeaders: {'Content-Type': 'application/json'}...)
```
I had to do this:
```
const DEFAULT_POST_HEADER: {[name: string]: string} = {
  'Content-Type': 'application/x-www-form-urlencoded'
};
NG2_UI_AUTH_PROVIDERS({defaultHeaders: DEFAULT_POST_HEADER...)
```

**Comment by ronzeidman** [comment]
Published as 3.0.2
Thanks for pointing out this issue!
